Tidy: detect ui tests subdirectory changes so `tests/ui/README.md` stays in sync

reddevilmidzy opened the pull request:
close: rust-lang/rust#150399 

There's an issue where `tests/ui/README.md` isn't updated whenever the ui subdirectory changes.
I've added subdirectory change detection to tidy ~~added a new mention to `triage.toml` to notify `tests/ui/README.md` to also be updated~~.

r? @Urgau

reddevilmidzy commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -73,6 +79,56 @@ pub fn check(root_path: &Path, tidy_ctx: TidyCtx) {
+            prev_line = line;
+        })
+        .map(|line| line.to_string())
+        .collect();
```
The logic to check if it is sorted is duplicated in issues.txt. Do I need to separate it into a function?

reddevilmidzy replied:
Yes, I refactored it.

Urgau commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -105,13 +161,30 @@ fn recursively_check_ui_tests<'issues>(
+        );
+        if !file_path.ends_with("ui/README.md") && !subdirectory.starts_with("ui-fulldeps/") {
+            cur_subdirs.insert(subdirectory);
+        }
```
Doing this string manipulation on every file seems wasteful to me, can't we read all the directories at once (without recursion)?

Maybe with something like this:
```rust
let entries = std::fs::read_dir(ui.as_path())?;

let cur_subdirs: Vec<String> = entries
        .filter_map(|entry| entry.ok())
        .map(|entry| entry.path())
        .filter(|path| path.is_dir())
        .map(|dir_path| {
            let dir_path = dir_path.strip_prefix(path).unwrap();
            dir_path.to_string_lossy().replace(std::path::MAIN_SEPARATOR_STR, "/")
        })
        .collect();
```

reddevilmidzy replied:
and also removed the trailing slashes from the directory strings in `ui_suites.txt`. Having the slashes at the end was making the sorting look a bit off.

Urgau commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -73,6 +79,56 @@ pub fn check(root_path: &Path, tidy_ctx: TidyCtx) {
+        let tidy_src = root_path.join("src/tools/tidy/src");
+        // instead of overwriting the file, recreate it and use an "atomic rename"
+        // so we don't bork things on panic or a contributor using Ctrl+C
+        let blessed_ui_suites_path = tidy_src.join("ui_suites.txt");
```
This should not be the same name, otherwise we'll still bork things.

```suggestion
        let blessed_ui_suites_path = tidy_src.join("ui_suites_blessed.txt");
```

Urgau commented:
Looks good to me, but I will let the final review to a T-bootstrap member.

*Btw, could you make the title and description a bit more descriptive.*

r? bootstrap

reddevilmidzy commented:
Okay, thank you for the review:)

Urgau commented:
> I don't quite understand the purpose of having a separate `ui_suites.txt`.
> 
> If the goal is to keep `tests/ui/README.md` in sync with the subdirectories, wouldn't it make more sense to compare the list of subdirectories to the list of headings in the readme?

Parsing a Markdown file is non-trivial and `tidy` currently doesn't have a dependency on [pulldown-cmark](https://lib.rs/crates/pulldown-cmark). I'm not sure we would want to add such dependency to `tidy`, so we would have to approximate it.

But in general yes, looking at `tests/ui/README.md` is probably better.

reddevilmidzy committed:
Add subdirectory change detection to tidy

Zalathar commented:
@rustbot author

reddevilmidzy commented:
Unified by removing '/' at the end of the title and added directories that did not have descriptions.

created:
```
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/annotate-moves` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/any` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/collections` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/deref` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/dist` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/eii` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/float` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/fn_traits` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/higher-ranked-trait-bounds` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/loop-match` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/offload` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/pin` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/pin-ergonomics` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/reborrow` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/rust-2018` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/rust-2021` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/rust-2024` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/sanitize-attr` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/scalable-vectors` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/sized-hierarchy` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/stack-probes` created
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/unstable-feature-bound` created
```

deleted
```
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/rust-{2018,2021,2024}` deleted
tidy [ui_tests (tests)]: ui subdirectory `tests/ui/typeof` deleted
```

thank you for the review:)
@rustbot ready

Zalathar commented on tests/ui/README.md:
```diff
@@ -113,45 +125,45 @@ See [Tracking Issue for autodiff #124509](https://github.com/rust-lang/rust/issu
 Tests for automatic referencing and dereferencing behavior, such as automatically adding reference operations (`&` or `&mut`) to make a value match a method's receiver type. Sometimes abbreviated as "auto-ref" or "auto-deref".
 
-## `tests/ui/backtrace/`: Backtraces
+## `tests/ui/backtrace`: Backtraces
```
I would prefer *not* to normalize the trailing-slash style in this PR, as I think it's a noisy change that's outside the primary purpose of these changes.

Instead, I think we should leave the existing headings as-is, and worry about backslash style in a follow-up PR.

(Of course, this means teaching the parser to ignore trailing slashes, which should be straightforward.)

Zalathar commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -73,6 +73,39 @@ pub fn check(root_path: &Path, tidy_ctx: TidyCtx) {
+            if let Some(start) = line.find('`')
+                && let Some(end) = line[start + 1..].find('`')
+            {
+                let title = &line[start + 1..start + 1 + end];
```
I think it's worth trying to write this parsing part as a regex, using `static_regex!` for memoization, and using a named capture group (e.g. ``(?<dir>[^`]*)``) to extract the directory name.

(Maybe it ends up being worse? But I think it's worth trying.)

Zalathar commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -73,6 +73,39 @@ pub fn check(root_path: &Path, tidy_ctx: TidyCtx) {
+    // Compare previous subdirectory with current subdirectory
+    // to sync with `tests/ui/README.md`.
+    // See <https://github.com/rust-lang/rust/issues/150399>
+    let prev_subdirs: BTreeSet<_> = include_str!("../../../../tests/ui/README.md")
```
Calling this `prev_subdirs` is technically true (assuming it previously passed CI), but feels a bit too indirect.

I think it would be clearer to name this after the fact that it contains the directory names extracted from the README.

Zalathar commented:
Sorry for the delay; I didn't see that this had been marked ready.

@rustbot author

reddevilmidzy committed:
Add section to tests/ui/README.md

reddevilmidzy commented on src/tools/tidy/src/ui_tests.rs:
```diff
@@ -73,6 +73,39 @@ pub fn check(root_path: &Path, tidy_ctx: TidyCtx) {
+            }
+        }
+        check.error("please update `tests/ui/README.md`");
+    }
```
Changed to a more specific description!

reddevilmidzy committed:
Reorder tests/ui/README.md

reddevilmidzy reviewed (commented):
Thank you for your review. I've applied your feedback.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150474/1b4325211c3228dd10eea12e9b3f749f87363d95..e103a20b61c37a7a064de09a833e80dcf317a027)*

Zalathar commented:
Nitpick: Could you please reorder the commits, so that the readme changes come *before* adding the tidy check?

That way, every individual commit should successfully pass tests, which is a nice property to have.

After that, I think we should be good to go.

@rustbot author

reddevilmidzy reviewed (commented):
Thank you for the review! @Zalathar 

@rustbot ready

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/150474/466ea4e6c39f8a43727edcc726ca86b499e14d83..a08541085aa3390ddf88339a2adeb29e8a118315)*

Zalathar commented:
Thanks for your patience. 🙌

@bors r+
